include/boost/geometry/algorithms/visit.hpp

```diff
@@ -0,0 +1,307 @@
+            }, geom1, geom2);
+    }
+};
+
+
+template <typename Geometry, typename Tag = typename tag<Geometry>::type>
+struct visit_breadth_first
+{
+    template <typename F, typename G>
+    static bool apply(F f, G & g)
+    {
+        return f(g);
+    }
+};
+
+template <typename Geometry>
+struct visit_breadth_first<Geometry, void>
+{
+    BOOST_GEOMETRY_STATIC_ASSERT_FALSE(
+        "Not implemented for this Geometry type.",
+        Geometry);
+};
+
+template <typename Geometry>
+struct visit_breadth_first<Geometry, dynamic_geometry_tag>
+{
+    template <typename Geom, typename F>
+    static bool apply(F function, Geom & geom)
+    {
+        bool result = true;
+        traits::visit<util::remove_cref_t<Geom>>::apply([&](auto & g)
+        {
+            result = visit_breadth_first<decltype(g)>::apply(function, g);
+        }, geom);
+        return result;
+    }
+};
+
+template <typename Geometry>
+struct visit_breadth_first<Geometry, geometry_collection_tag>
+{
+    template <typename F, typename Geom>
+    static bool apply(F function, Geom & geom)
+    {
+        using iter_t = typename boost::range_iterator<Geom>::type;
+        std::deque<iter_t> queue;
+
+        iter_t it = boost::begin(geom);
+        iter_t end = boost::end(geom);
+        for(;;)
+        {
+            for (; it != end; ++it)
+            {
+                bool result = true;
+                traits::visit_iterator<util::remove_cref_t<Geom>>::apply([&](auto & g)
+                {
+                    result = visit_or_enqueue(function, g, queue, it);
+                }, it);
+                if (! result)
+                    return false;
+            }
+
+            if (queue.empty())
+            {
+                break;
+            }
+
+            // Alternatively store a pointer to GeometryCollection
+            // so this call can be avoided.
+            traits::visit_iterator<util::remove_cref_t<Geom>>::apply([&](auto & g)
+            {
+                set_iterators(g, it, end);
+            }, queue.front());
+            queue.pop_front();
+        }
+
+        return true;
+    }
+
+private:
+    template
+    <
+        typename F, typename Geom, typename Iterator,
+        std::enable_if_t<util::is_geometry_collection<Geom>::value, int> = 0
+    >
+    static bool visit_or_enqueue(F &, Geom &, std::deque<Iterator> & queue, Iterator iter)
+    {
+        queue.push_back(iter);
+        return true;
+    }
+    template
+    <
+        typename F, typename Geom, typename Iterator,
+        std::enable_if_t<! util::is_geometry_collection<Geom>::value, int> = 0
+    >
+    static bool visit_or_enqueue(F & f, Geom & g, std::deque<Iterator> & , Iterator)
+    {
+        return f(g);
+    }
+
+    template
+    <
+        typename Geom, typename Iterator,
+        std::enable_if_t<util::is_geometry_collection<Geom>::value, int> = 0
+    >
+    static void set_iterators(Geom & g, Iterator & first, Iterator & last)
```

Comment:
I guess it could not be `Geom const& g`

Reply:
The point is to support both const and non-const references here. This is also why a function template argument is used instead of struct template argument `Geometry`.